File: helpers/configure_git.py
import configparser
import logging
import os
import re
import shlex
import subprocess
import sys
from pathlib import Path

from rich.console import Console
from rich.logging import RichHandler
from rich.prompt import Confirm, Prompt
from rich.text import Text

# --- Logger and Global Setup ---
log = logging.getLogger("bankai.git_config")
console = Console()
log.setLevel(logging.INFO)
log.propagate = False
log.addHandler(
    RichHandler(
        console=console,
        rich_tracebacks=True,
        show_path=False,
        show_level=False,
        show_time=False,
    )
)


# --- Helper Functions ---
def run_command(command, check=True):
    """Runs a shell command interactively, as this script requires user input."""
    log.info(f"[dim]Executing: {command}[/dim]")
    try:
        # These commands may need interactive prompts (e.g., ssh-add passphrase)
        subprocess.run(command, shell=True, check=check)
    except subprocess.CalledProcessError as e:
        log.error(f"Command '{command}' failed with exit code {e.returncode}.")
    except FileNotFoundError:
        log.error(f"Command not found for: '{command}'")


def start_ssh_agent():
    """Starts ssh-agent and sets the environment variables for the current process."""
    log.info("Starting ssh-agent...")
    try:
        result = subprocess.run(
            ["ssh-agent", "-s"], capture_output=True, text=True, check=True
        )
        # Parse the output of ssh-agent -s to set environment variables
        find_vars = re.compile(r"(\S+)=([^;]+);")
        for line in result.stdout.strip().split("\n"):
            match = find_vars.search(line)
            if match:
                key, value = match.groups()
                os.environ[key] = value
        # Log the agent PID for confirmation
        if "Agent pid" in result.stdout:
            log.info(result.stdout.strip().split("\n")[-1])
    except (subprocess.CalledProcessError, FileNotFoundError) as e:
        log.error(
            f"Could not start ssh-agent. Please ensure it's installed. Error: {e}"
        )
        sys.exit(1)


def prompt_user(prompt_text, default=False):
    """Asks the user a yes/no question, handling Ctrl+C gracefully."""
    try:
        return Confirm.ask(prompt_text, default=default)
    except KeyboardInterrupt:
        log.warning("\nOperation cancelled by user.")
        sys.exit(1)


# --- Main Logic ---
def create_profile(profile_type: str, ssh_dir: Path):
    """Creates a complete Git profile: directory, .gitconfig, and SSH key."""
    log.info(f"--- Setting up {profile_type.capitalize()} Git profile ---")
    profile_dir = Path.home() / profile_type
    profile_dir.mkdir(exist_ok=True)

    email = Prompt.ask(f"Enter {profile_type} email")
    username = Prompt.ask(f"Enter {profile_type} username")
    github_user = Prompt.ask(f"Enter GitHub username for {profile_type}")

    key_path = ssh_dir / f"{profile_type}_key"
    gitconfig_path = profile_dir / f".gitconfig.{profile_type}"

    # Use configparser for a structured approach
    config = configparser.ConfigParser()
    config["user"] = {
        "email": email,
        "name": username,
        "signingkey": str(key_path),
    }
    config["github"] = {"user": f'"{github_user}"'}
    config["commit"] = {"gpgsign": "true"}
    config["gpg"] = {"format": "ssh"}
    config["core"] = {"sshCommand": f'"ssh -i {key_path}"'}

    with gitconfig_path.open("w") as f:
        config.write(f)
    log.info(f"Created {gitconfig_path}")

    log.info(f"Generating SSH key for {profile_type}...")
    run_command(f'ssh-keygen -t ed25519 -C "{email}" -f {key_path} -N "" -q')

    log.info(f"Adding {profile_type} SSH key to agent...")
    run_command(f"ssh-add {key_path}")


def main():
    """Main execution flow for the Git configuration script."""
    log.info("Configuring Git...")
    ssh_dir = Path.home() / ".ssh"
    ssh_dir.mkdir(mode=0o700, exist_ok=True)

    # Start SSH agent
    start_ssh_agent()

    work_profile_created = False
    if prompt_user("Do you want to create a work Git profile?", default=True):
        create_profile("work", ssh_dir)
        work_profile_created = True

    create_profile("personal", ssh_dir)

    # Use configparser for the global gitconfig as well
    global_config = configparser.ConfigParser()
    global_config['includeIf "gitdir:~/personal/"'] = {
        "path": "~/personal/.gitconfig.personal"
    }
    if work_profile_created:
        global_config['includeIf "gitdir:~/work/"'] = {"path": "~/work/.gitconfig.work"}

    global_gitconfig_path = Path.home() / ".gitconfig"
    with global_gitconfig_path.open("w") as f:
        global_config.write(f)
    log.info(f"Created global git config at {global_gitconfig_path}")
    log.info(
        Text.from_markup(
            "[bold yellow]ACTION REQUIRED:[/bold yellow] Copy the contents of the .pub files in ~/.ssh and add them to your GitHub accounts."
        )
    )


if __name__ == "__main__":
    main()

File: src/helpers/configure_git.js
import fs from "node:fs";
import { homedir, tmpdir } from "node:os";
import path from "node:path";
import promptsLib from "prompts";
import {
	log,
	promptUser as promptUserDefault,
	safeCopyFile,
} from "../common/utils.js";

const HOME = homedir();
const GIT_PROFILES = Object.freeze({
	personal: {
		directory: "dev",
		filename: ".gitconfig.dev",
		legacyDirectory: "personal",
		legacyFilename: ".gitconfig.personal",
	},
	work: {
		directory: "Work",
		filename: ".gitconfig.work",
		legacyDirectory: "work",
		legacyFilename: ".gitconfig.work",
	},
});

function migrateLegacyProfiles(home) {
	const availableProfiles = new Set();
	for (const [profileType, profile] of Object.entries(GIT_PROFILES)) {
		const legacyDirectory = path.join(home, profile.legacyDirectory);
		const legacyProfile = path.join(legacyDirectory, profile.legacyFilename);
		const destination = path.join(home, profile.directory, profile.filename);
		if (
			profileType === "work" &&
			fs.existsSync(legacyProfile) &&
			!fs.existsSync(destination)
		) {
			fs.mkdirSync(path.dirname(destination), { recursive: true });
			fs.renameSync(legacyProfile, destination);
			log.info(`Moved legacy Git profile ${legacyProfile} to ${destination}.`);
		}
		if (fs.existsSync(legacyProfile)) {
			log.warning(
				`Legacy Git profile found at ${legacyProfile}; move it to ${destination}.`,
			);
		}
		if (fs.existsSync(destination)) availableProfiles.add(profileType);
	}
	return availableProfiles;
}

/**
 * Default runner: spawns an argv array via Bun.spawn and returns its exit code.
 *
 * Using an argv array (not a shell string) is the whole point — the free-form
 * email from the prompt is passed as a discrete argument to ssh-keygen, so a
 * quote or `;` in the email can never break out into shell interpretation.
 *
 * @param {string[]} argv
 * @returns {Promise<{ exitCode: number }>}
 */
async function defaultRunner(argv) {
	const proc = Bun.spawn(argv, {
		stdin: "inherit",
		stdout: "inherit",
		stderr: "inherit",
	});
	const exitCode = await proc.exited;
	return { exitCode };
}

/** Default ssh-agent bootstrap (no-op-able in tests via opts.startAgent). */
async function startSshAgent() {
	try {
		const agentProc = Bun.spawn(["ssh-agent", "-s"]);
		const output = await new Response(agentProc.stdout).text();
		const match = output.match(/SSH_AUTH_SOCK=([^;]+);/);
		if (match) {
			process.env.SSH_AUTH_SOCK = match[1];
		}
		const pidMatch = output.match(/SSH_AGENT_PID=([^;]+);/);
		if (pidMatch) {
			process.env.SSH_AGENT_PID = pidMatch[1];
		}
	} catch (_e) {
		log.warning("Could not start ssh-agent.");
	}
}

/**
 * Create a single Git profile (email/name/signing key) under ~/dev or ~/Work.
 *
 * Key generation is guarded two ways:
 *   - If the key already exists, generation is SKIPPED (ssh-keygen would
 *     otherwise hang on an interactive "Overwrite (y/n)?" prompt) but the
 *     signing config is still emitted — we keep using the existing key.
 *   - If generation runs and FAILS, the profile gitconfig is written WITHOUT
 *     any signing directives. Emitting `gpgsign = true` + a `signingkey` that
 *     doesn't exist would make every future commit in that gitdir fail to sign.
 *
 * @param {string} profileType e.g. "personal" | "work"
 * @param {{ home: string, sshDir: string, promptFn: Function, runner: Function }} ctx
 */
async function createProfile(profileType, { home, sshDir, promptFn, runner }) {
	log.info(`--- Setting up ${profileType} Git profile ---`);
	const profile = GIT_PROFILES[profileType];
	const profileDir = path.join(home, profile.directory);
	fs.mkdirSync(profileDir, { recursive: true });

	const response = await promptFn([
		{
			type: "text",
			name: "email",
			message: `Enter ${profileType} email`,
		},
		{
			type: "text",
			name: "username",
			message: `Enter ${profileType} username`,
		},
		{
			type: "text",
			name: "githubUser",
			message: `Enter GitHub username for ${profileType}`,
		},
	]);

	if (!response.email || !response.username || !response.githubUser) {
		log.error("Missing information. Skipping profile creation.");
		return;
	}

	const keyPath = path.join(sshDir, `${profileType}_key`);
	const gitConfigPath = path.join(profileDir, profile.filename);

	// `signingOk` decides whether we emit the SSH-signing directives. It stays
	// true when we reuse an existing key or generate one successfully, and flips
	// to false only when a fresh generation fails.
	let signingOk = true;

	if (fs.existsSync(keyPath)) {
		// Skip generation: ssh-keygen would otherwise stop on an interactive
		// overwrite prompt and hang the whole setup. The existing key is reused.
		log.info(
			`SSH key ${keyPath} already exists — skipping generation, reusing it.`,
		);
	} else {
		log.info(`Generating SSH key for ${profileType}...`);
		// argv array, NOT a shell string — the email is a discrete argument so
		// quotes/semicolons in it can never be interpreted by a shell.
		const { exitCode } = await runner([
			"ssh-keygen",
			"-t",
			"ed25519",
			"-C",
			response.email,
			"-f",
			keyPath,
			"-N",
			"",
			"-q",
		]);
		if (exitCode !== 0) {
			signingOk = false;
			log.warning(
				`ssh-keygen failed (exit ${exitCode}) for ${profileType} — writing gitconfig WITHOUT SSH signing so future commits aren't blocked by a missing key.`,
			);
		} else {
			// Add to agent (best-effort; argv array, no shell).
			log.info(`Adding ${profileType} SSH key to agent...`);
			await runner(["ssh-add", keyPath]);
		}
	}

	// Build the profile gitconfig. Signing directives are appended only when a
	// usable key is present.
	let gitConfigContent = `[user]
    email = ${response.email}
    name = ${response.username}`;
	if (signingOk) {
		gitConfigContent += `
    signingkey = ${keyPath}`;
	}
	gitConfigContent += `

[github]
    user = "${response.githubUser}"
`;
	if (signingOk) {
		gitConfigContent += `
[commit]
    gpgsign = true

[gpg]
    format = ssh

[core]
    sshCommand = "ssh -i ${keyPath}"
`;
	}

	fs.writeFileSync(gitConfigPath, gitConfigContent);
	log.info(`Created ${gitConfigPath}`);
}

/**
 * Configure Git profiles + the global ~/.gitconfig includeIf routing.
 *
 * Global-config overwrite semantics come from utils.safeCopyFile.
 *
 * Every collaborator is injectable for tests (defaults preserve production
 * behavior): `home`, `sshDir`, `promptFn` (the prompts() multi-question lib),
 * `promptUser` (the yes/no confirm from utils — Ctrl+C aborts, handled there),
 * `runner` (Bun.spawn argv wrapper), and `startAgent`.
 *
 * @param {{
 *   home?: string,
 *   sshDir?: string,
 *   promptFn?: Function,
 *   promptUser?: Function,
 *   runner?: (argv: string[]) => Promise<{ exitCode: number }>,
 *   startAgent?: () => Promise<void>,
 * }} [opts]
 */
export async function configureGit(opts = {}) {
	const {
		home = HOME,
		sshDir = path.join(home, ".ssh"),
		promptFn = promptsLib,
		promptUser = promptUserDefault,
		runner = defaultRunner,
		startAgent = startSshAgent,
	} = opts;

	log.info("Configuring Git...");
	const availableProfiles = migrateLegacyProfiles(home);
	fs.mkdirSync(sshDir, { mode: 0o700, recursive: true });

	await startAgent();

	const ctx = { home, sshDir, promptFn, runner };

	let workProfileCreated = availableProfiles.has("work");
	if (await promptUser("Do you want to create a work Git profile?", true)) {
		await createProfile("work", ctx);
		workProfileCreated = true;
	}

	await createProfile("personal", ctx);

	// Global Config
	const personalProfile = GIT_PROFILES.personal;
	let globalConfigContent = `[includeIf "gitdir:~/${personalProfile.directory}/"]
    path = ~/${personalProfile.directory}/${personalProfile.filename}
`;

	if (workProfileCreated) {
		const workProfile = GIT_PROFILES.work;
		globalConfigContent += `[includeIf "gitdir:~/${workProfile.directory}/"]
    path = ~/${workProfile.directory}/${workProfile.filename}
`;
	}

	const globalGitConfigPath = path.join(home, ".gitconfig");

	// safeCopyFile takes a source path, so stage the generated content outside
	// the home tree where it cannot pollute ~/.ssh.
	const stagePath = path.join(
		fs.mkdtempSync(path.join(tmpdir(), "haoshoku-gitcfg-")),
		"gitconfig",
	);
	fs.writeFileSync(stagePath, globalConfigContent);
	try {
		const wrote = safeCopyFile(stagePath, globalGitConfigPath);
		if (wrote) {
			log.info(`Created global git config at ${globalGitConfigPath}`);
		} else {
			log.info(
				`Global git config at ${globalGitConfigPath} already up to date — left untouched.`,
			);
		}
	} finally {
		fs.rmSync(path.dirname(stagePath), { recursive: true, force: true });
	}

	log.warning(
		"ACTION REQUIRED: Copy the contents of the .pub files in ~/.ssh and add them to your GitHub accounts.",
	);
}
